main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,524 @@
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation._
+
+/**
+  * The Pattern Exhaustiveness phase checks pattern matches for exhaustiveness
+  *
+  * A pattern match is exhaustive if there exists a type correct expression which is not matched by any of the patterns
+  * in the pattern match
+  *
+  * Implemented as described in
+  * MARANGET, L. (2007). Warnings for pattern matching. Journal of Functional Programming, 17(3), 387-421. doi:10.1017/S0956796807006223
+  *
+  * The algorithm works by recursively "peeling off layers" of the pattern match. It check if the outermost type
+  * constructors are exhaustive, then if so, recursively calls itself on the arguments. If it finds a pattern which
+  * is not exhaustive during a recursive case, it uses it to built up a pattern that is not matched by the full
+  * pattern match and returns the result.
+  *
+  */
+object PatternExhaustiveness extends Phase[TypedAst.Root, TypedAst.Root] {
+
+  sealed trait TypeConstructor
+  object TypeConstructor {
+    case class True() extends TypeConstructor
+    case class False() extends TypeConstructor
+    case class Unit() extends TypeConstructor
+    case class Tuple(args: List[TypeConstructor]) extends TypeConstructor
+    case class BigInt() extends TypeConstructor
+    case class Int8() extends TypeConstructor
+    case class Int16() extends TypeConstructor
+    case class Int32() extends TypeConstructor
+    case class Int64() extends TypeConstructor
+    case class Float32() extends TypeConstructor
+    case class Float64() extends TypeConstructor
+    case class Str() extends TypeConstructor
+    case class Char() extends TypeConstructor
+    case class Wild() extends TypeConstructor
+    case class Enum(name: String, sym: EnumSym, numArgs: Int, args: List[TypeConstructor]) extends TypeConstructor
+  }
+
+  /**
+    * A small ADT to track if we've found a non exhaustive pattern
+    *
+    * Essentially a reverse Optional: we only include a parameter if there is an error
+    */
+  sealed trait Exhaustiveness
+  object Exhaustiveness {
+    case class Exhaustive() extends Exhaustiveness
+    case class NonExhaustive(pat: List[TypeConstructor]) extends Exhaustiveness
+  }
+
+  /**
+    * Returns an error message if a pattern match is not exhaustive
+    */
+  def run(root: TypedAst.Root)(implicit flix: Flix): Validation[TypedAst.Root, CompilationError] = {
+    implicit val _ = flix.genSym
+    val startTime = System.nanoTime()
+
+    for {
+      defns <- seqM(root.definitions.map { case (_, v) => Definitions.checkPats(root, v) })
+    } yield {
+      val currentTime = System.nanoTime()
+      val time = root.time.copy(typer = currentTime - startTime)
+      root
+    }
+  }
+
+  object Definitions {
+    def checkPats(root: TypedAst.Root, tast: TypedAst.Declaration.BoundedLattice)(implicit genSym: GenSym): Validation[TypedAst.Declaration.BoundedLattice, CompilationError] = tast match {
```

Comment:
We usually keep the root parameter as the last parameter. The reasoning is that the more important the parameter is the closer to the start it should be.